File: jvm/core/src/main/java/se/idkollen/client/endpoints/FtnEndpoint.java
package se.idkollen.client.endpoints;

import se.idkollen.client.PollOptions;
import se.idkollen.client.internal.Transport;

import se.idkollen.client.models.*;

import java.util.concurrent.CompletableFuture;

/** Endpoint for Finnish Trust Network (FTN) operations. */
public class FtnEndpoint {
    private final Transport transport;

    public FtnEndpoint(Transport transport) {
        this.transport = transport;
    }

    /** Start an FTN authentication session. */
    public CompletableFuture<FtnStatus> auth(FtnAuthRequest req) {
        return transport.post("/v3/ftn/auth", req, FtnStatus.class);
    }

    /** Start an FTN age verification session. */
    public CompletableFuture<AgeVerificationStatus> ageVerification(AgeVerificationRequest req) {
        return transport.post("/v3/ftn/age-verification", req, AgeVerificationStatus.class);
    }

    /** Poll the current status of an FTN authentication session. */
    public CompletableFuture<FtnStatus> authStatus(String id) {
        return transport.get("/v3/ftn/auth/" + id, FtnStatus.class);
    }

    /** Poll the current status of an FTN age verification session. */
    public CompletableFuture<AgeVerificationStatus> ageVerificationStatus(String id) {
        return transport.get("/v3/ftn/age-verification/" + id, AgeVerificationStatus.class);
    }

    /** Cancel an FTN authentication session. */
    public CompletableFuture<Void> cancelAuth(String id) {
        return transport.delete("/v3/ftn/auth/" + id);
    }

    /** Cancel an FTN age verification session. */
    public CompletableFuture<Void> cancelAgeVerification(String id) {
        return transport.delete("/v3/ftn/age-verification/" + id);
    }

    /** Poll until the authentication session reaches a terminal state or the timeout elapses. */
    public CompletableFuture<FtnStatus> waitForAuth(String id, PollOptions opts) {
        return transport.poll(() -> authStatus(id), FtnStatus.Pending.class::isInstance, opts);
    }

    public CompletableFuture<FtnStatus> waitForAuth(String id) {
        return waitForAuth(id, new PollOptions());
    }

    /** Poll until the age verification session reaches a terminal state or the timeout elapses. */
    public CompletableFuture<AgeVerificationStatus> waitForAgeVerification(String id, PollOptions opts) {
        return transport.poll(() -> ageVerificationStatus(id), AgeVerificationStatus.Pending.class::isInstance, opts);
    }

    public CompletableFuture<AgeVerificationStatus> waitForAgeVerification(String id) {
        return waitForAgeVerification(id, new PollOptions());
    }
}

File: php/src/Endpoints/FtnEndpoint.php
<?php

declare(strict_types=1);

namespace Idkollen\Client\Endpoints;

use Idkollen\Client\Models\AgeVerification\AgeVerificationCompleted;
use Idkollen\Client\Models\AgeVerification\AgeVerificationFailed;
use Idkollen\Client\Models\AgeVerification\AgeVerificationPending;
use Idkollen\Client\Models\AgeVerification\AgeVerificationRequest;
use Idkollen\Client\Models\AgeVerification\AgeVerificationStatus;
use Idkollen\Client\Models\Ftn\FtnAuthRequest;
use Idkollen\Client\Models\Ftn\FtnCompleted;
use Idkollen\Client\Models\Ftn\FtnFailed;
use Idkollen\Client\Models\Ftn\FtnPending;
use Idkollen\Client\Models\Ftn\FtnStatus;
use Idkollen\Client\PollOptions;
use Idkollen\Client\Transport;
use Idkollen\Client\WaitException;

final class FtnEndpoint
{
    public function __construct(private readonly Transport $transport) {}

    public function auth(FtnAuthRequest $req): FtnStatus
    {
        return self::parseStatus($this->transport->post('/v3/ftn/auth', self::serialize($req)));
    }

    public function ageVerification(AgeVerificationRequest $req): AgeVerificationStatus
    {
        return self::parseAgeStatus($this->transport->post('/v3/ftn/age-verification', self::serialize($req)));
    }

    public function authStatus(string $id): FtnStatus
    {
        return self::parseStatus($this->transport->get('/v3/ftn/auth/' . $id));
    }

    public function ageVerificationStatus(string $id): AgeVerificationStatus
    {
        return self::parseAgeStatus($this->transport->get('/v3/ftn/age-verification/' . $id));
    }

    public function cancelAuth(string $id): void
    {
        $this->transport->delete('/v3/ftn/auth/' . $id);
    }

    public function cancelAgeVerification(string $id): void
    {
        $this->transport->delete('/v3/ftn/age-verification/' . $id);
    }

    public function waitForAuth(string $id, PollOptions $opts = new PollOptions()): FtnStatus
    {
        $deadline = time() + $opts->timeout;
        while (true) {
            $status = $this->authStatus($id);
            if (!$status instanceof FtnPending) {
                return $status;
            }
            if (time() >= $deadline) {
                throw new WaitException(timeout: true);
            }
            sleep($opts->interval);
        }
    }

    public function waitForAgeVerification(string $id, PollOptions $opts = new PollOptions()): AgeVerificationStatus
    {
        $deadline = time() + $opts->timeout;
        while (true) {
            $status = $this->ageVerificationStatus($id);
            if (!$status instanceof AgeVerificationPending) {
                return $status;
            }
            if (time() >= $deadline) {
                throw new WaitException(timeout: true);
            }
            sleep($opts->interval);
        }
    }

    private static function serialize(object $req): array
    {
        return array_filter(get_object_vars($req), fn($v) => $v !== null);
    }

    private static function parseStatus(array $d): FtnStatus
    {
        return match ($d['status']) {
            'PENDING' => new FtnPending(
                id: $d['id'],
                refId: $d['refId'] ?? null,
                url: $d['url'],
            ),
            'COMPLETED' => new FtnCompleted(
                id: $d['id'],
                refId: $d['refId'] ?? null,
                ssn: $d['ssn'],
                name: $d['name'],
                givenName: $d['givenName'],
                surname: $d['surname'],
                phone: $d['phone'] ?? null,
                email: $d['email'] ?? null,
                address: $d['address'] ?? null,
                birthDate: $d['birthDate'] ?? null,
                pid: $d['pid'] ?? null,
                bankId: $d['bankId'] ?? null,
            ),
            'FAILED' => new FtnFailed(
                id: $d['id'],
                refId: $d['refId'] ?? null,
                error: $d['error'],
            ),
            default => throw new \UnexpectedValueException("Unknown ftn status: {$d['status']}"),
        };
    }

    private static function parseAgeStatus(array $d): AgeVerificationStatus
    {
        return match ($d['status']) {
            'PENDING' => new AgeVerificationPending(
                id: $d['id'],
                url: $d['url'] ?? null,
                minAge: $d['minAge'] ?? null,
                maxAge: $d['maxAge'] ?? null,
            ),
            'COMPLETED' => new AgeVerificationCompleted(
                id: $d['id'],
                ageVerified: $d['ageVerified'],
            ),
            'FAILED' => new AgeVerificationFailed(
                id: $d['id'],
                error: $d['error'],
            ),
            default => throw new \UnexpectedValueException("Unknown age verification status: {$d['status']}"),
        };
    }
}
